agent/xds/listeners_ingress.go

```diff
@@ -284,33 +347,92 @@ func makeSDSOverrideFilterChains(cfgSnap *proxycfg.ConfigSnapshot,
 	return chains, nil
 }
 
-func makeCommonTLSContextFromSDS(sdsCfg structs.GatewayTLSSDSConfig) *envoy_tls_v3.CommonTlsContext {
+var EnvoyTLSVersions = map[types.TLSVersion]envoy_tls_v3.TlsParameters_TlsProtocol{
+	types.TLSVersionAuto: envoy_tls_v3.TlsParameters_TLS_AUTO,
+	types.TLSv1_0:        envoy_tls_v3.TlsParameters_TLSv1_0,
+	types.TLSv1_1:        envoy_tls_v3.TlsParameters_TLSv1_1,
+	types.TLSv1_2:        envoy_tls_v3.TlsParameters_TLSv1_2,
+	types.TLSv1_3:        envoy_tls_v3.TlsParameters_TLSv1_3,
+}
+
+func EnvoyTLSCipherSuites(cipherSuites []types.TLSCipherSuite) []string {
+	cipherSuiteStrings := []string{}
+	for _, c := range cipherSuites {
+		cipherSuiteStrings = append(cipherSuiteStrings, types.EnvoyTLSCipherSuiteStrings[c])
+	}
+	return cipherSuiteStrings
+}
+
+// TODO: could these two functions become a single function that takes an interface as an argument?
+func makeTLSParametersFromGatewayTLSConfig(tlsCfg structs.GatewayTLSConfig) *envoy_tls_v3.TlsParameters {
+	tlsParams := envoy_tls_v3.TlsParameters{}
+
+	if tlsCfg.TLSMinVersion != types.TLSVersionInvalid {
```

Comment:
So this usage confuses me. It looks like the zero-value of the `TLSVersion` type was [changed to be `TLSVersionInvalid`](https://github.com/hashicorp/consul/pull/11645/files#diff-8aef003676b31aa3e5e3fea823c827812f96e4ef0155a8ac582147a5393eeea8R9). But, if we just bypass setting the versions here, the problem is that envoy's zero-values for these fields according to the [generated proto code](https://github.com/envoyproxy/go-control-plane/blob/17d64b8311fdba876a3c2a259c3a28e388647296/envoy/api/v2/auth/common.pb.go#L34) and the [docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto) is to effectively have the same behavior as explicitly setting (in our case) `TLSVersionAuto`.

Shouldn't we be throwing an error if `types.TLSVersionInvalid` is set rather than defaulting to the zero-value behavior?

Also, if that's the case and we do return an error for `TLSVersionInvalid` shouldn't the zero-value we set be equivalent to the zero-value in envoy (rather than an invalid value)? Otherwise aren't we introducing a breaking change?

Reply:
What i discovered through writing tests was that `TLSVersionAuto` is not a good zero-value, because when a gateway defines a min TLS version and a listener does not specify a min version, the zero value from the listener struct would actually unset, on the listener, the min version from the gateway.

My thinking wass that it should be possible to unset a gateway default by explicitly passing `TLS_AUTO` in listener config, but a listener shouldn't unset a gateway default when TLS config is unspecified.

Added `TLSVersionUnspecified` as a distinct valid zero-value over in #11645 and updated these checks to use that instead.